magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,678 @@
+    seconds = (midi_ticks / constants.STANDARD_PPQ) \
+               * self.state.seconds_per_beat
+    self.state.time_position -= seconds
+
+  def parsedirection(self, xml_direction):
+    """Parse the MusicXML <direction> element."""
+
+    for child in xml_direction:
+      if child.tag == "sound":
+        if child.get("tempo") != None:
+          tempo = Tempo(self.state, child)
+          self.tempos.append(tempo)
+          self.state.tempo = tempo.qpm
+          self.state.seconds_per_beat = 60 / self.state.tempo
+          if child.get("dynamics") != None:
+            self.state.velocity = int(child.get("dynamics"))
+
+  def parseforward(self, xml_forward):
+    """Parse the MusicXML <backup> element.
+    This moves the global time position forward.
+    """
+
+    xml_duration = xml_forward.find('duration')
+    forward_duration = int(xml_duration.text)
+    midi_ticks = forward_duration * (constants.STANDARD_PPQ
+                                     / self.state.divisions)
+    seconds = (midi_ticks / constants.STANDARD_PPQ) \
+               * self.state.seconds_per_beat
+    self.state.time_position += seconds
+
+
+class Note(object):
+  """Internal representation of a MusicXML <note> element"""
+  def __init__(self, xml_note, state):
+    self.xml_note = xml_note
+    self.voice = 1
+    self.is_rest = False
+    self.is_in_chord = False
+    self.is_grace_note = False
+    self.pitch = None               # Tuple (Pitch Name, MIDI number)
+    self.note_duration = NoteDuration(state)
+    self.state = state
+    self.parse()
+
+  def parse(self):
+    """Parse the MusicXML <note> element"""
+
+    self.midi_channel = self.state.midi_channel
+    self.midi_program = self.state.midi_program
+    self.velocity = self.state.velocity
+
+    for child in self.xml_note:
+      if child.tag == "chord":
+        self.is_in_chord = True
+      elif child.tag == "duration":
+        self.note_duration.parseduration(self, child.text)
+      elif child.tag == "pitch":
+        self.parsepitch(child)
+      elif child.tag == "rest":
+        self.is_rest = True
+      elif child.tag == "voice":
+        self.voice = int(child.text)
+      elif child.tag == "dot":
+        self.note_duration.dots += 1
+      elif child.tag == "type":
+        self.note_duration.type = child.text
+      elif child.tag == "time-modification":
+        # A time-modification element represents a tuplet_ratio
+        self.parsetuplet(child)
+
+  def parsepitch(self, xml_pitch):
+    """Parse the MusicXML <pitch> element"""
+    step = xml_pitch.find("step").text
+    alter = 0
+    if xml_pitch.find("alter") != None:
+      alter = xml_pitch.find("alter").text
+    octave = xml_pitch.find("octave").text
+
+    # Visual pitch representation
+    alter_string = ""
+    if alter == "-2":
+      alter_string = "bb"
+    elif alter == "-1":
+      alter_string = "b"
+    elif alter == "1":
+      alter_string = "#"
+    elif alter == "2":
```

Comment:
For this one, maybe add an else branch that logs a message at the `debug` level, along with a comment that we're currently ignoring decimal semitones. We've thought about adding support for that sort of thing in the (likely distant) future, so it would be good to have an obvious place to go back and add it.